### PR TITLE
Iskbaga/tiger rebased new

### PR DESCRIPTION
**Pull request by iskbaga**

### CODE REVIEW

File: configs/train/letter.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
-    "experiment_name": "letter_data",
-    "best_metric": "validation/ndcg@20",
+    "experiment_name": "letter_tiger",
+    "best_metric": "loss",
```
Comment (NonameUntitled) [inline configs/train/letter.json:3]: А почему тут лосс?

File: modeling/loss/base.py [hunk 2]
```diff
@@ -141,7 +141,21 @@ def forward(self, inputs):
             inputs[self._output_prefix] = loss.cpu().item()
 
         return loss
-
```
Comment (NonameUntitled) [inline modeling/loss/base.py:144]: Отступ 2 строки

File: modeling/loss/base.py [hunk 3]
```diff
@@ -141,7 +141,21 @@ def forward(self, inputs):
...
+        self._input_loss_key = predictions_prefix
+        self._output_prefix = output_prefix
+
+    def forward(self, inputs):
+        loss = inputs[self._input_loss_key]
+        assert loss.dim() == 0, "Loss must be a scalar tensor"
+        if self._output_prefix is not None:
+            inputs[self._output_prefix] = loss.cpu().item()
+        return loss
+
```
Comment (NonameUntitled) [inline modeling/loss/base.py:158]: Отступ 2 строки

File: modeling/utils/NewTigerTest.py
Comment (NonameUntitled) [inline modeling/utils/NewTigerTest.py:1]: Создай папочку tests рядом с modeling

File: modeling/models/tiger.py [hunk 4]
```diff
@@ -1,489 +1,337 @@
...
+                    tgt=tgt,
+                    memory=after_encoder_emb,
+                    tgt_mask=tgt_mask,
+                    memory_key_padding_mask=memory_key_padding_mask
+                )  # (batch_size, L, embedding_dim)
 
-        codebook_embeddings = self._get_position_embeddings(
-            lengths, mask, codebook_lambda, self._codebook_embeddings
-        )
+                last_output = decoder_output[:, -1:, :]  # (batch_size, 1, embedding_dim)
```
Comment (NonameUntitled) [inline modeling/models/tiger.py:273]: тут можно сделать `decoder_output[:, -1, :]`

File: modeling/models/tiger.py [hunk 5]
```diff
@@ -1,489 +1,337 @@
...
 
-        return position_embeddings + codebook_embeddings
+                weights = self.codebook_embeddings.weight[step * self._codebook_size: (step + 1) * self._codebook_size]
+                logits = torch.matmul(
+                    last_output,
+                    weights.t()
+                ).squeeze(1)  # (batch_size, codebook_size)
```
Comment (NonameUntitled) [inline modeling/models/tiger.py:279]: Если сделаешь то, что я выше написал, то можно тут squeeze не делать

**Review by NonameUntitled — CHANGES_REQUESTED** [review]
